ocaml/xapi/xapi_xenvmd.ml

```diff
@@ -0,0 +1,123 @@
+(* 
+ * Code to manage xenvmd
+ *
+ * This is a little bit of a layering violation.
+ *
+ *)
+
+module D=Debug.Make(struct let name="xapi_xenvmd" end)
+open D
+
+let rpc x =
+  if !Xcp_client.use_switch
+  then Xcp_client.json_switch_rpc !(Storage_interface.queue_name) x
+  else Xcp_client.http_rpc Xmlrpc.string_of_call Xmlrpc.response_of_string ~srcstr:"xapi" ~dststr:"xenops" Storage_interface.uri x
+			  
+(* We need to be able to talk SMAPIv2 *)
+let local_url () = Http.Url.(Http { host="127.0.0.1"; auth=None; port=None; ssl=false }, { uri = "/services/SM"; query_params=["pool_secret",!Xapi_globs.pool_secret] } )
+module SM = Storage_interface.Client(struct let rpc = rpc end)
+
+exception NoConfigFile
+exception XenvmdFailed
+
+(* The storage backends know about these paths *)
+let lockfile sr = Printf.sprintf "/var/lib/xenvmd/%s.lock" sr
+let configfile sr = Printf.sprintf "/etc/xenvm.d/VG_XenStorage-%s.xenvmd.config" sr
+
+let is_running sr =
+  debug "Checking for xenvmd running for sr: %s" sr;
+  try
+    let l = lockfile sr in
+    ignore(Unix.stat l);
+    let fd = Unix.openfile l [ Unix.O_WRONLY; Unix.O_CREAT ] 0o0644 in
+    Pervasiveext.finally
+      (fun () -> Unix.lockf fd Unix.F_TEST 1)
+      (fun () -> Unix.close fd);
+    debug "Locked successfully: xenvmd not running";
+    false
+  with
+  | Unix.Unix_error (Unix.ENOENT, _, _) -> debug "Caught ENOENT (xenvmd not running)"; false (* Lockfile missing *)
+  | Unix.Unix_error (Unix.EAGAIN, _, _) -> debug "Caught EAGAIN (xenvmd running)"; true  (* Locked by xenvmd *)
+  | Unix.Unix_error (Unix.EACCES, _, _) -> debug "Caught EACCESS (xenvmd running)"; true  (* Locked by xenvmd *)
+  | e -> raise e
+
+let assert_config_present sr =
+  try
+    let l = configfile sr in
+    ignore(Unix.stat l);
+  with
+  | Unix.Unix_error (Unix.ENOENT, _, _) -> raise NoConfigFile
+  | e -> raise e
+           
+let really_start sr =
+  assert_config_present sr;
+  try
+    let output, stderr = Forkhelpers.execute_command_get_output !Xapi_globs.xenvmd_path
+        [ "--daemon"; "--config"; configfile sr ] in
+    (* Note that xenvmd will detect multiple copies of itself for the same VG.
+       it will write an error message to the stdout/stderr but exit with code
+       0 *)
+    debug "Xenvmd started for SR %s. output='%s' stderr='%s'" sr output stderr
+  with
+  | Forkhelpers.Spawn_internal_error(log, output, x) ->
+    debug "Failed to execute xenvmd. log='%s' output='%s'" log output;
+    raise XenvmdFailed
```

Comment:
Should we put some of the debug information from the failed spawn into the exception?